arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -354,24 +349,23 @@ public fun <A> resource(block: suspend ResourceScope.() -> A): Resource<A> = blo
  * <!--- KNIT example-resource-06.kt -->
  */
 @ScopeDSL
-public suspend fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
+public suspend inline fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
   val scope = ResourceScopeImpl()
-  val a: A = try {
+  var finished = false
+  return try {
     action(scope)
   } catch (e: Throwable) {
-    val ex = if (e is CancellationException) ExitCase.Cancelled(e) else ExitCase.Failure(e)
-    val ee = withContext(NonCancellable) {
-      scope.cancelAll(ex, e) ?: e
+    finished = true
+    scope.cancelAll(ExitCase.ExitCase(e))
```

Comment:
`cancelAll` deals with `NonCancellable` internally now, which is good because it can now throw the errors itself. By that I mean that the code:
```
withContext(NonCancellable) {
  throw someExceptionWithSuppressed
}
```
actually strips the suppressed exceptions due to internal coroutine machinery, and so we before had to throw the exception outside, but now we encapsulate all that logic in `closeAll`, which also ensures consistency in semantics between this and `allocated`